### PR TITLE
[core] Add event to notify when RenderSource has been loaded

### PR DESCRIPTION
**Pull request by jmalanen**
Resolves: mapbox/mapbox-gl-native-team#82

### CODE REVIEW

File: src/mbgl/map/map_impl.cpp [hunk 1]
```diff
@@ -154,7 +154,7 @@ void Map::Impl::onWillStartRenderingMap() {
 }
 
 void Map::Impl::onDidFinishRenderingMap() {
-    if (mode == MapMode::Continuous && loading) {
```
Comment (pozdnyakov) [inline src/mbgl/map/map_impl.cpp]: why is this change required?

File: src/mbgl/renderer/render_source.cpp [hunk 2]
```diff
@@ -64,12 +64,39 @@ void RenderSource::setObserver(RenderSourceObserver* observer_) {
...
+        }
+        observer->onIdealTilesLoaded(*this);
+    }
 }
 
 void RenderSource::onTileError(Tile& tile, std::exception_ptr error) {
     observer->onTileError(*this, tile.id, error);
 }
 
+void RenderSource::onIdealTilesChanged(const std::vector<UnwrappedTileID>& idealTiles) {
```
Comment (pozdnyakov) [inline src/mbgl/renderer/render_source.cpp]: maybe it would be simpler, if we just introduced `const std::set<UnwrappedTileID>& TilePyramid::getIdealTiles() const `

File: src/mbgl/renderer/render_source.hpp [hunk 3]
```diff
@@ -118,6 +118,10 @@ class RenderSource : protected TileObserver {
 
     void onTileChanged(Tile&) override;
     void onTileError(Tile&, std::exception_ptr) final;
+    void onIdealTilesChanged(const std::vector<UnwrappedTileID>&) final;
+
+private:
+    std::unordered_map<UnwrappedTileID, bool> idealTilesLoaded;
```
Comment (pozdnyakov) [inline src/mbgl/renderer/render_source.hpp]: instead, it could be `std::set<UnwrappedTileID> idealTilesLoaded;` that would hold only loaded tiles

File: src/mbgl/renderer/render_source.cpp [hunk 4]
```diff
@@ -64,12 +64,39 @@ void RenderSource::setObserver(RenderSourceObserver* observer_) {
 
 void RenderSource::onTileChanged(Tile& tile) {
     observer->onTileChanged(*this, tile.id);
+    if (tile.isComplete()) {
+        const auto tileID = tile.id.toUnwrapped();
+        const auto& idealTile = idealTilesLoaded.find(tileID);
+        if (idealTile != idealTilesLoaded.end()) {
+            idealTilesLoaded[tileID] = true;
+        }
+        for (const auto& entry : idealTilesLoaded) {
```
Comment (pozdnyakov) [inline src/mbgl/renderer/render_source.cpp]: here, we could just compare the size of `idealTilesLoaded` with the size of the actual ideal tiles from the tile pyramid (pls see the comments below)

File: src/mbgl/renderer/render_source.cpp [hunk 5]
```diff
@@ -64,12 +64,39 @@ void RenderSource::setObserver(RenderSourceObserver* observer_) {
 
 void RenderSource::onTileChanged(Tile& tile) {
     observer->onTileChanged(*this, tile.id);
+    if (tile.isComplete()) {
+        const auto tileID = tile.id.toUnwrapped();
+        const auto& idealTile = idealTilesLoaded.find(tileID);
```
Comment (pozdnyakov) [inline src/mbgl/renderer/render_source.cpp]:
```
if (tilePyramid.getIdealTiles().count(tileID)) {
idealTilesLoaded.insert(tileID);
}
```

File: src/mbgl/map/map_impl.cpp (hunk 1 above)
Comment (jmalanen) [reply src/mbgl/map/map_impl.cpp]: I removed `loading` because it is checked again two lines below. Having the check here prevents sending `onDidFinishRenderingMap` event again after loading new tiles caused by panning or zooming the map.

File: src/mbgl/renderer/render_source.cpp (hunk 5 above)
Comment (jmalanen) [reply src/mbgl/renderer/render_source.cpp]: Couple issues here. RenderSource does not have visibility to tilePyramid. Did you mean` if (tilePyramid.getIdealTiles().count(tileID) == 0) `? If the tile already exists,  `insert` will not do anything.

File: src/mbgl/renderer/render_source.cpp (hunk 2 above)
Comment (jmalanen) [reply src/mbgl/renderer/render_source.cpp]: Done

File: src/mbgl/renderer/render_source.hpp (hunk 3 above)
Comment (jmalanen) [reply src/mbgl/renderer/render_source.hpp]: Done

File: src/mbgl/renderer/render_source.cpp (hunk 4 above)
Comment (jmalanen) [reply src/mbgl/renderer/render_source.cpp]: Done
